### PR TITLE
[SP-6300] SSH Step Refactor

### PR DESCRIPTION
**Pull request by igorkheif**

### CODE REVIEW

**Comment by hagaishapira** [comment]
@igorkheif is this ready for CR?

**Comment by igorkheif** [comment]
@hagaishapira not yet

File: steps/ssh/parallel-command/ssh_parallel_command.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+type SSHParallelArgs struct {
+	sshBase.SSHArgs
+	HostNames []string `env:"HOSTNAMES,required" envDefault:""`
+}
+
+type ParallelSSHCommand struct {
+	args SSHParallelArgs
+}
+
+func ParseArgs() (SSHParallelArgs, step.Status) {
```
Comment (hagaishapira) [inline steps/ssh/parallel-command/ssh_parallel_command.go:34]: I don't think you need this function, you can just call env.Parse from Init() and then do the validation if status isn't Failed()

File: steps/ssh/command/manifest.yaml [hunk 2]
```diff
@@ -49,6 +49,31 @@ envs:
...
+  - name: GCLOUD_PROJECT
+    description: "GCP project where the secret is sotred. Must be provided if GCP_SECRET_KEY is used."
+    type: "string"
+    required: false
+    example: "77/zQpdlq3oUoB2P66Lu/3ZxgE3JoQA5vqPtV7dA"
+  - name: AUTH_CODE
+    description: "Base64 encoded key token file. Must be provided if GCP_SECRET_KEY is used and no GCP integration is provided."
+    type: "string"
+    required: false
+    example: "77/zQpdlq3oUoB2P66Lu/3ZxgE3JoQA5vqPtV7dA"
```
Comment (hagaishapira) [inline steps/ssh/command/manifest.yaml:76]: Please fix output - if it's string then it should be just a string and not a json

File: steps/ssh/base/ssh.go [hunk 3]
```diff
@@ -1,93 +1,160 @@
...
+		return status.New(codes.InvalidArgument, "aws region is required when specifyng aws secret", "")
 	}
 
-	return result, nil
+	// GCP
+	if args.GCPSecretKey != "" && args.GCPAuthCode == "" {
+		return status.New(codes.InvalidArgument, "either a GCP integration or AUTH_CODE are required when specifying gcp secret", "")
+	}
+	if args.GCPSecretKey != "" && args.GCPProject == "" {
+		return status.New(codes.InvalidArgument, "gcp project is required when specifyng gcp secret", "")
```
Comment (hagaishapira) [inline steps/ssh/base/ssh.go:88]: please be consistent about case when writing GCP and AWS, you can always use upper case but search for all instances in the code and align

File: steps/ssh/base/ssh.go [hunk 4]
```diff
@@ -1,93 +1,160 @@
...
 	StrictHostKeyChecking string        `env:"STRICT_HOST_KEY_CHECKING" envDefault:"no"`
 	LogLevel              string        `env:"LOG_LEVEL" envDefault:"ERROR"`
 	Port                  int           `env:"PORT" envDefault:"22"`
 	ConnectionTimeout     time.Duration `env:"CONNECTION_TIMEOUT" envDefault:"30s"`
 }
 
-func ParseArgs() (SSHArgs, error) {
+func ParseArgs() (SSHArgs, step.Status) {
 	result := SSHArgs{}
 	err := envconf.Parse(&result)
```
Comment (hagaishapira) [inline steps/ssh/base/ssh.go:54]: don't use envconf directly use out wrapped env.Parse()

File: steps/ssh/base/ssh.go [hunk 5]
```diff
@@ -1,93 +1,160 @@
...
+	"github.com/stackpulse/steps-sdk-go/step"
+
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/secretsmanager"
 
+	gcpSecretmanager "cloud.google.com/go/secretmanager/apiv1"
+	gcpSecretmanagerpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1"
+
 	envconf "github.com/caarlos0/env/v6"
```
Comment (hagaishapira) [inline steps/ssh/base/ssh.go:23]: don't need this

File: steps/ssh/parallel-command/ssh_parallel_command.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/stackpulse/steps-sdk-go/codes"
+	"github.com/stackpulse/steps-sdk-go/status"
+	sshBase "github.com/stackpulse/steps/ssh/base"
+
+	"sync"
+
+	"github.com/stackpulse/steps-sdk-go/step"
+)
+
+type SSHResponse struct {
```
Comment (hagaishapira) [inline steps/ssh/parallel-command/ssh_parallel_command.go:16]: rename this to SSHOutput and please make sure to have json tags which use snake case

File: steps/ssh/parallel-command/ssh_parallel_command.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	Hostname string
+	Success  bool
+	Output   string
+	Error    string
```
Comment (hagaishapira) [inline steps/ssh/parallel-command/ssh_parallel_command.go:20]:
```suggestion
	Hostname string `json:"hostname"`
	Success  bool `json:"success"`
	Output   string `json:"output"`
	Error    string `json:"error"`
```

File: steps/ssh/parallel-command/manifest.yaml [hunk 8]
```diff
@@ -45,20 +85,21 @@ envs:
...
+    description: "Timeout waiting for SSH connection (default 30s)."
     type: "string"
     required: false
-    example: "us-west-2"
+    example: "1m"
+
 
 outputs:
-  - name: SSHResponses
+  - name: output
```
Comment (hagaishapira) [inline steps/ssh/parallel-command/manifest.yaml:102]: do we know if this step is being used anywhere as we are having non-BC changes

File: steps/ssh/parallel-command/manifest.yaml [hunk 9]
```diff
@@ -45,20 +85,21 @@ envs:
...
     description: "JSON array contains the ssh command results for each host"
     type: "json"
     example: "[{\"Hostname\":\"192.168.125.110\",\"Success\":true,\"Output\":\"ssh-command\\n\",\"Error\":\"\"},{\"Hostname\":\"192.168.125.112\",\"Success\":true,\"Output\":\"ssh-command\\n\",\"Error\":\"\"}]"
```
Comment (hagaishapira) [inline steps/ssh/parallel-command/manifest.yaml:105]: adjust output to new struct member names

File: steps/ssh/parallel-command/manifest.yaml [hunk 10]
```diff
@@ -15,9 +15,9 @@ envs:
     example: "ubuntu"
   - name: HOSTNAMES
     description: "Hostnames for ssh command execution. Max hosts allowed - 1000."
-    type: "json"
+    type: "string"
```
Comment (hagaishapira) [inline steps/ssh/parallel-command/manifest.yaml:18]: please add in the description that if a command fails on any host the step will return success status with the error listed under the failed host in the output.

**Review by hagaishapira — CHANGES_REQUESTED** [review]
Overall looks much better, see comments and please also delete the duplicate steps with the specific cloud providers in this PR

File: steps/ssh/parallel-command/ssh_parallel_command.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	HostNames []string `env:"HOSTNAMES,required" envDefault:""`
+}
+
+type ParallelSSHCommand struct {
+	args SSHParallelArgs
+}
+
+func ParseArgs() (SSHParallelArgs, step.Status) {
+	result := SSHParallelArgs{}
+	err := envconf.Parse(&result)
```
Comment (oriser) [inline steps/ssh/parallel-command/ssh_parallel_command.go:36]: The same here with the env

**Review by oriser — CHANGES_REQUESTED** [review]
See comments

File: steps/ssh/parallel-command/ssh_parallel_command.go (hunk 1 above)
Comment (igorkheif) [reply steps/ssh/parallel-command/ssh_parallel_command.go]: 👍

File: steps/ssh/command/manifest.yaml (hunk 2 above)
Comment (igorkheif) [reply steps/ssh/command/manifest.yaml:76]: 👍

File: steps/ssh/base/ssh.go (hunk 5 above)
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

File: steps/ssh/base/ssh.go (hunk 4 above)
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

File: steps/ssh/base/ssh.go (hunk 3 above)
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

File: steps/ssh/parallel-command/ssh_parallel_command.go (hunk 6 above)
Comment (igorkheif) [reply steps/ssh/parallel-command/ssh_parallel_command.go:14]: `SSHOutput` is the new one, left `SSHResponse` as the BC one

File: steps/ssh/parallel-command/ssh_parallel_command.go (hunk 7 above)
Comment (igorkheif) [reply steps/ssh/parallel-command/ssh_parallel_command.go:18]: 👍

File: steps/ssh/parallel-command/manifest.yaml (hunk 8 above)
Comment (igorkheif) [reply steps/ssh/parallel-command/manifest.yaml:102]: 👍

File: steps/ssh/parallel-command/manifest.yaml (hunk 9 above)
Comment (igorkheif) [reply steps/ssh/parallel-command/manifest.yaml]: 👍

File: steps/ssh/parallel-command/manifest.yaml (hunk 10 above)
Comment (igorkheif) [reply steps/ssh/parallel-command/manifest.yaml:18]: 👍

File: steps/ssh/base/ssh.go [hunk 12]
```diff
@@ -1,93 +1,160 @@
 package base
 
 import (
+	"context"
+	"encoding/base64"
 	"fmt"
 	"io/ioutil"
+	"os"
 	"strconv"
 	"time"
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/base/ssh.go:12]: 👍

File: steps/ssh/base/ssh.go [hunk 13]
```diff
@@ -1,93 +1,160 @@
...
 
+	"github.com/stackpulse/steps-sdk-go/codes"
+	"github.com/stackpulse/steps-sdk-go/status"
+	"github.com/stackpulse/steps-sdk-go/step"
+
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/secretsmanager"
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

File: steps/ssh/base/ssh.go [hunk 14]
```diff
@@ -1,93 +1,160 @@
...
-	AWSRegion             string        `env:"AWS_REGION" envDefault:""`
 	PrivateKey            string        `env:"PRIVATE_KEY" envDefault:""`
 	Password              string        `env:"PASSWORD" envDefault:""`
+	AWSSecretKey          string        `env:"AWS_SECRET_KEY" envDefault:""`
+	AWSRegion             string        `env:"AWS_REGION" envDefault:""`
+	AWSAccessKeyId        string        `env:"AWS_ACCESS_KEY_ID,required" envDefault:""`
+	AWSSecretAccessKey    string        `env:"AWS_SECRET_ACCESS_KEY,required" envDefault:""`
+	GCPSecretKey          string        `env:"GCP_SECRET_KEY" envDefault:""`
+	GCPProject            string        `env:"GCLOUD_PROJECT" envDefault:""`
+	GCPAuthCode           string        `env:"AUTH_CODE" envDefault:""`
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

File: steps/ssh/base/ssh.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

File: steps/ssh/base/ssh.go [hunk 15]
```diff
@@ -1,93 +1,160 @@
...
+			return "", nil, status.FromError(codes.InvalidArgument, "failed fetching aws secret", err)
+		}
+	}
+	if s.GCPSecretKey != "" {
+		var err error
+		privateKey, err = s.FetchGcpSecret()
+		if err != nil {
+			return "", nil, status.FromError(codes.InvalidArgument, "failed fetching gcp secret", err)
 		}
 	}
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

File: steps/ssh/base/ssh.go [hunk 16]
```diff
@@ -1,93 +1,160 @@
...
 
 	err := ioutil.WriteFile(PrivateKeyPath, []byte(privateKey), 0644)
 	if err != nil {
-		return "", nil, fmt.Errorf("write private key: %w", err)
+		return "", nil, step.Status{
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

File: steps/ssh/base/ssh.go [hunk 17]
```diff
@@ -1,93 +1,160 @@
...
+			Message: "encountered an internal error",
+			Verbose: fmt.Sprintf("write private key: %v", err),
+		}
 	}
 
 	// Restrict key.pem file capabilities (for ssh usage)
 	output, _, err := sdkExec.Execute("chmod", []string{"600", PrivateKeyPath})
 	if err != nil {
-		return "", nil, fmt.Errorf("chmod private key: %w: %s", err, output)
+		return "", nil, step.Status{
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

File: steps/ssh/command/ssh_command.go [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+	sshCmd, sshArgs, st = s.args.BuildCommand(s.args.Hostname)
+	if st.Failed() {
+		return nil, st
+	}
+
+	rawOutput, st := sshBase.ExecuteSSHCommand(sshCmd, sshArgs)
+
+	return step.Output{
+		"output": string(rawOutput),
+	}, st
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/command/ssh_command.go:35]: 👍

File: steps/ssh/parallel-command/ssh_parallel_command.go (hunk 11 above)
Comment (igorkheif) [reply steps/ssh/parallel-command/ssh_parallel_command.go]: 👍

File: steps/ssh/base/ssh.go [hunk 19]
```diff
@@ -1,93 +1,160 @@
...
 	result := SSHArgs{}
 	err := envconf.Parse(&result)
 	if err != nil {
-		return result, err
+		return result, status.FromError(codes.InvalidArgument, "failed parsing arguments", err)
+	}
+	return result, status.OK
+}
+
+func ValidateArgs(args SSHArgs) step.Status {
```
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/base/ssh.go:57]: 👍

File: steps/ssh/parallel-command/manifest.yaml (hunk 8 above)
Comment (igorkheif) [reply steps/ssh/parallel-command/manifest.yaml:102]: Now returning new style and old style for BC as previously discussed

File: steps/ssh/command/ssh_command.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/command/ssh_command.go:35]: This has always been "output". The one over at "ssh/parallel-command" is fixed.

**Review by hagaishapira — APPROVED** [review]
Really nice work! Have @oriser take a look and merge away

File: steps/ssh/base/ssh.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (igorkheif) [reply steps/ssh/base/ssh.go]: 👍

**Review by oriser — APPROVED** [review]
Looks great

**Pull request merged**